py_ecc/BaseCurve.py

```diff
@@ -0,0 +1,291 @@
+from abc import (
+    abstractmethod,
+)
+
+from typing import (
+    cast,
+)
+
+from py_ecc.field_elements import (
+    FQP,
+)
+
+from py_ecc.optimized_field_elements import (
+    FQP as optimized_FQP,
+)
+
+from py_ecc.new_typing import (
+    Field,
+    GeneralPoint,
+    Point2D,
+)
+from py_ecc.new_typing import (
+    Optimized_Field,
+    Optimized_Point2D,
+    Optimized_Point3D,
+)
+
+
+class BaseCurve:
+    # Name of the curve can be "bn128" or "bls12_381"
+    curve_name = None
+    curve_order = None
+    # Curve is y**2 = x**3 + b
+    b = None
+    # Twisted curve over FQ**2
+    b2 = None
+    # Extension curve over FQ**12; same b value as over FQ
+    b12 = None
+    # Generator for curve over FQ
+    G1 = None
+    # Generator for twisted curve over FQ2
+    G2 = None
+    # Generator for twisted curve over FQ12
+    G12 = None
+    # Point at infinity over FQ
+    Z1 = None
+    # Point at infinity for twisted curve over FQ2
+    Z2 = None
+
+    def __init__(self, curve_properties, curve_name):
+        self.curve_name = curve_name
+        self.curve_order = curve_properties[curve_name]["curve_order"]
```

Comment:
It appears that we only access `curve_properties[curve_name]` which suggests that we could modify the `__init__` argument to just take that value so that these lookups become:

```python
self.b = curve_properties["b"]
self.b2 = curve_properties["b"]
```